### PR TITLE
Derive Show with field labels

### PR DESCRIPTION
**Pull request by TimWSpence**

### CODE REVIEW

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+object show extends ShowDerivation
+
+trait ShowDerivation:
+
+  extension (F: Show.type)
+    inline def derived[A](using gen: K0.Generic[A]): Show[A] = gen.derive(productShow, coproductShow)
+
+  given productShow[A](using inst : => K0.ProductInstances[Show, A], l: Labelling[A]): Show[A] with
+    def show(a: A) = {
```
Comment (joroKr21) [inline core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:14]: Let's use a SAM function instead (it doesn't generate anonymous classes)

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      var idx = 0
+      val x = inst.foldLeft[StringBuilder](a)(new StringBuilder(s"${l.label}("))(
+        [t] => (acc: StringBuilder, sh: Show[t], t0: t) => {
+          //elemLabels is backed by an array so this should be fast
+          val res = Continue(acc.append(s"${l.elemLabels(idx)}=${sh.show(t0)}, "))
+          idx = idx + 1
+          res
+          }
+      )
+      x.delete(x.length - 2, x.length).append(")")
```
Comment (joroKr21) [inline core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:24]: I think this is not safe for empty products

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      x.toString()
+    }
+
+  given coproductShow[A](using inst: => K0.CoproductInstances[Show, A]): Show[A] with
+    def show(a: A) = inst.fold[String](a)(
```
Comment (joroKr21) [inline core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:29]: Same here - let's use a SAM function

**Review by joroKr21 — CHANGES_REQUESTED** [review]
Really nice 👍 See my comments for some minor issues.

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala (hunk 2 above)
Comment (TimWSpence) [reply core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:24]: 🤦 good point!

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala (hunk 1 above)
Comment (TimWSpence) [reply core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:14]: You mean define an explicit `ProductShow` trait like we've done for the other typeclasses?

Comment (joroKr21) [reply core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:14]: I mean you can just use a function:
`given ... : Show[A] = a => ...`

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      val x = inst.foldLeft[StringBuilder](a)(new StringBuilder(s"${l.label}("))(
+        [t] => (acc: StringBuilder, sh: Show[t], t0: t) => {
+          //elemLabels is backed by an array so this should be fast
+          val res = Continue(acc.append(s"${l.elemLabels(idx)}=${sh.show(t0)}, "))
+          removeTrailingComma = true
+          idx = idx + 1
+          res
+          }
+      )
+      if(removeTrailingComma) {x.delete(x.length - 2, x.length)}
```
Comment (joroKr21) [inline core/src/main/scala-3.0.0-RC3/cats/derived/show.scala:26]: Isn't this just
```suggestion
      if (i > 0) {x.delete(x.length - 2, x.length)}
```

File: core/src/main/scala-3.0.0-RC3/cats/derived/show.scala (hunk 2 above)
Comment (joroKr21) [reply core/src/main/scala-3.0.0-RC3/cats/derived/show.scala]: Isn't this just
```suggestion
      if (idx > 0) {x.delete(x.length - 2, x.length)}
```

Comment (TimWSpence) [reply core/src/main/scala-3.0.0-RC3/cats/derived/show.scala]: I should just give up on programming today 😂 Thanks

**Review by joroKr21 — APPROVED** [review]
Perfect 👍

**Pull request merged**